contracts/e2e-test/E2E.sol

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2025 RISC Zero, Inc.
+//
+// All rights reserved.
+
+pragma solidity ^0.8.9;
+
+import {Test} from "forge-std/Test.sol";
+import {console2} from "forge-std/console2.sol";
+import {Vm} from "forge-std/Vm.sol";
+import {MessageHashUtils} from "@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol";
+import {IRiscZeroVerifier} from "risc0/IRiscZeroVerifier.sol";
+import {RiscZeroSetVerifier} from "risc0/RiscZeroSetVerifier.sol";
+import {RiscZeroVerifierRouter} from "risc0/RiscZeroVerifierRouter.sol";
+import {RiscZeroCheats} from "risc0/test/RiscZeroCheats.sol";
+import {IRiscZeroSelectable} from "risc0/IRiscZeroSelectable.sol";
+import {UnsafeUpgrades, Upgrades} from "openzeppelin-foundry-upgrades/Upgrades.sol";
+
+import {Account} from "../src/types/Account.sol";
+import {AssessorJournal} from "../src/types/AssessorJournal.sol";
+import {AssessorReceipt} from "../src/types/AssessorReceipt.sol";
+import {Callback} from "../src/types/Callback.sol";
+import {Fulfillment} from "../src/types/Fulfillment.sol";
+import {Input, InputType} from "../src/types/Input.sol";
+import {Requirements} from "../src/types/Requirements.sol";
+import {Offer} from "../src/types/Offer.sol";
+import {ProofRequest} from "../src/types/ProofRequest.sol";
+import {Predicate, PredicateType} from "../src/types/Predicate.sol";
+import {RequestId, RequestIdLibrary} from "../src/types/RequestId.sol";
+import {RequestLock} from "../src/types/RequestLock.sol";
+
+import {IBoundlessMarket} from "../src/IBoundlessMarket.sol";
+import {BoundlessMarket} from "../src/BoundlessMarket.sol";
+import {BoundlessMarketLib} from "../src/libraries/BoundlessMarketLib.sol";
+import {HitPoints} from "../src/HitPoints.sol";
+import {IHitPoints} from "../src/HitPoints.sol";
+
+Vm constant VM = Vm(0x7109709ECfa91a80626fF3989D68f67F5b1DD12D);
+bytes32 constant APP_IMAGE_ID = 0x722705a82a1dab8369b17e16bac42c9c538057fc1d32933d21ea2b47f292efb4;
+uint256 constant DEFAULT_BALANCE = 1000 ether;
+
+/// Test designed to checks that a full e2e flow works.
+contract E2E is RiscZeroCheats, Test {
```

Comment:
Cool that we can test things e2e in Solidity!

What I'm unclear on is when we should write e2e tests here vs in E2E tests in Boundless Market. Should we always do both? Or are there particular types of features we should cover here and not in the market

Reply:
Since we have the Rust e2e tests that include the Router, Groth16, etc, I'm personally inclined to drop this one. Foundry FFI is fragile and requires pre-installing the CLI (see GH actions config) and I think the Rust version of this test is likely to run faster / be a better foundation for adding more variations.